### ISSUE TITLE
"Components currently only support simple events - you cannot include arguments. Sorry!"

### ISSUE THREAD

**Issue by dagnelies** [ractivejs/ractive#1684]
Hi,

I was just curious if there was anything planned in the short or long term to support normal events for components.

Cheers,
Arnaud

**Comment by Rich-Harris** [comment]
Nothing particular, though this is a weak spot for sure. The trouble is that element events and component events are fundamentally different but look identical...

```html
<!-- fires on DOM click event, which has predictable structure -->
<button on-click='doSomething()'>click me</button>

<!-- fires on synthetic 'select' event, which could have x arguments -->
<fancy-list on-select='doSomething()' items='{{myItems}}'/>
```

...which makes predictable composition somewhat tricky. You can imagine the inside of the `<fancy-list>` component looking something like this...

```js
{{#each items}}
  <button on-click='fire("select",this)'>select {{this.name}}</button>
{{/each}}
```

...in which case how is the `on-select` handler meant to know which item has been selected? At the moment the [solution is to simply rename the event](http://jsfiddle.net/rich_harris/pontdky5/) and pass it along unchanged, which requires that the `on-select` directive has no arguments, and doesn't invoke a method.

But then *that* means that you can't differentiate between components as effectively:

```html
{{#each lists :i}}
  <!-- nope, can't do this -->
  <fancy-list on-select='doSomething:{{i}}' items='{{this}}'/>
{{/each}}
```

Meanwhile, you can't have DOM event listeners on components...

```html
<fancy-list on-mouseover='doSomethingElse()' .../>
```

...which would often be useful.

So in short, there are (at least) three subtly different things we might want to do with `on-someevent` in components, and I for one haven't been able to figure out how to accommodate them all without straining the syntax to the point where its usage becomes hard to understand. I would welcome any ideas from the hive mind!

One thing that has changed since component events were first introduced is that it's possible to capture bubbling events from deep within your tree with `ractive.on('*.someevent',handler)`, which makes the use case that is currently prioritised (propagating events upwards) less of an issue.

**Comment by martypdx** [comment]
IMO, the biggest need is to get to the context in which the component is being used, as you clearly show in your example. Strangely enough, you can get the index via `this.event` but not if you use `fire()` and the context is still the "inside the component" context (see http://jsfiddle.net/rz84w7rr/):

```html
{{#each list:i}}
    <fancy-list on-select='select' item='{{this}}'/>
{{/each}}
<p>selected index: {{selected}}</p>
```
```js
r.on('select', function(){
    this.set('selected', this.event.index.i);
})
```

I agree that with event bubbling, the need for complexity in component tag defined events has been greatly diminished.  Perhaps biggest bang for the buck is to shore up the `this.event` to return context and any indexes correctly?

**Comment by dagnelies** [comment]
Hi,

thanks for the prompt replies, as usual ;)
...to be honest, I couldn't quite follow the entirety of the argument. I didn't even knew you could fire custom events until now. ;) 

What follows is a bit off-topic, more general discussion/opinion.

I currently find ractive fantastic. It does an amazing job to keep up the data binding and its straightforward to use. For me, this simplicity is the key, that's what I love about ractive and I think it is its greatest strength.

That works great for data-binding. And it would be nice to keep the rest that simple. One area where I often stumble upon, recurrently, is dealing with "functions&events", especially when in combination with components.

In the end, in the small mind of mine, I would like to just be able to declare a function like:
```
<script>
function whatever( arg ) {
     // ....
}
</script>
```
And call it "normally" in ractive. Be it as an expression or an event-handler, in a native element or a component. Everyone could then place the functions quick and dirty in gIobal scope, or in packages, or wherever. I don't know if there are any technical limitations or if this was a design decision, but from my own point of view, it would be easier if things were handled like regular functions. My 2 cents.

The one nice thing though are the "method calls" like "set('x',3)" which are very handy. I wouldn't like to loose that. So, dunno if it's possible and sounds good to you:
- First check the current ractive instance if there is such a method, exactly like now
- If not, call the one in global scope


Going one step further, I would also put event proxies and method calls under the same hood, like:
```
methodCall(a,b,c) -> methodCall(a,b,c) // stays the same
eventProxy:a,b,c -> eventProxy(event,a,b,c) // more homogenous, isn't it?
``` 
Like in usual javascript/html, you could pass over `event` and `this`.

If you look closer, the changes aren't that ground-breaking. It's:
- defaulting to global scope for function calls
- deprecating event proxies in favor of being able to pass the `event` argument to method calls

...however, I thinkit would have a great impact on usability. Simpler, more homogenous and noob friendly if you ask me.

But already like it's now, thumbs up guys!

**Comment by justspamjustin** [comment]
I'm not fully in favor of allowing event handlers to call functions on the global scope. I like it being scoped to the context of the template.
However, I will say that when I ran into this issue that component events are treated quite differently, it stumped me for a while. DOM events aside, I would still like to be able to call a method in the template so I don't have to go through the boilerplate of manually creating a listener for that event.

**Comment by JonDum** [comment]
> Meanwhile, you can't have DOM event listeners on components...
> ` <fancy-list on-mouseover='doSomethingElse()' .../>`

This this this this. 1000x times this. Need this so bad to be able to make meaningful, reusable elements. A button that displays a progress bar, a loading animation, etc.

It would also be super cool if components could 'extend' native elements a la Polymer.

**Comment by JonDum** [comment]
For example, I do something like this a lot:

```
            <button on-click='delete()' class='red mls'>
                {{{deleting ? '<div class="loading-ani></div>"' : 'DELETE'}}}
            </button>
```

But I really need to be able to call another component inside the unary (because my loading animations are complex as shit haha):

```
            <button on-click='delete()' class='red mls'>
                {{{deleting ? '<loading>' : 'DELETE'}}}
            </button>
```

Doesn't work for obvious reasons


So I have to do something like this every time!
```
            <button on-click='approve()' class='green'>
                {{#approving}} 
                    <loading/>
                {{else}}
                  'Approve'
                {{/}}
            </button>
```
Getting super verbose.

I'd love to be able to make a reusable component like this:

```
            <progress-button class='green'
                             spinon='{{approving}}'
                             on-click='approve()'>
                           Approve</progress-button>
```

With the super huge bonus of being able to change it once and change all the places it's used throughout the site!

**Comment by martypdx** [comment]
@JonDum agree it makes sense to allow method invocation from component event handlers. 

In addition, I think being able to reference and call methods on other components would be useful. In a broad sense this could be allowing chaining in method calls:

```html
<button on-click'findComponent('dialog').open()'>do something</button>
<dialog on-close='doCloseThing()' on-cancel-'doCancelThing()'/>
```

As a more limited step, or merely syntax sugar as I suspect that calling components would be common task, we could use the `$` which would have logical connection to it's historic jquery usage IMO:

```html
<button on-click'$dialog.open()'>do something</button>
```

> So I have to do something like this every time!

            <button on-click='approve()' class='green'>
                {{#approving}} 
                    <loading/>
                {{else}}
                  'Approve'
                {{/}}
            </button>
> Getting super verbose.

Have to disagree here, comparing them formatted the same way:

```html
{{{approving ? '<loading>' : 'Approve'}}}
{{#approving}}<loading/>{{else}}'Approve'{{/}}
```

There's not a big difference. :)

**Comment by guilhermeaiolfi** [comment]
How much work would it need? Is that something we can expect for ractive 0.8 (in case all core developers agree that it's the way to go)?

**Comment by JonDum** [comment]
> Have to disagree here, comparing them formatted the same way:

@martypdx Yea true. Not that bad when formatted that way.


Here's a question. What if when ractive comes across an `on-` DOM event attr on a component, it attaches event listener's to the root elements of the components template? Then in onteardown removes them for you? 

`special-button` component template:
```html
<button class="{{class}}" style="{{style}}">  
    Click me
</button>
```

Usage:
```html
<special-button class="large" on-click='doSomething()'>
</special-button>
```

I can't think of a scenario where you would want the component to dispatch a DOM event from inside a nested child and have it bubble to the component's parent... all while using the same name of the DOM event, so it would be fine to reserve the DOM event names for DOM events only on components, right?

**Comment by fskreuz** [comment]
@martypdx 

> In addition, I think being able to reference and call methods on other components would be useful. In a broad sense this could be allowing chaining in method calls
> ```
> <button on-click='findComponent('dialog').open()'>do something</button>
<dialog on-close='doCloseThing()' on-cancel-'doCancelThing()'/>
> ```

Agreed. Removes at least 3 lines already in the component definition.

```
openDialog: function(){
  this.findComponent('dialog').open();
}
```

> As a more limited step, or merely syntax sugar as I suspect that calling components would be common task, we could use the $ which would have logical connection to it's historic jquery usage IMO:
>```
><button on-click'$dialog.open()'>do something</button>
>```

NOOO!!!

**Comment by pembeci** [comment]
```
<special-button class="large" on-click='doSomething()'>
</special-button>
```
I can see how this will be useful but what will happen if the template of special-button contains more than one clickable HTML elements? Do you want to catch and handle all of them at the parent? Will it be wise to force this as the default behavior for any type of component?

May be what we need is having a catalog of the different types of component events use cases. Then, we can look at them all together and try to come up with a way to accomodate each in a consistent way and/or prioritize some of them (so the most useful ones will have a more favorable/concise syntax).

**Comment by martypdx** [comment]
@pembeci they are events fired by the component, not DOM level events. So it's up to the component author. 

Could either be in code or passing on an event:

    <button on-click='click'>button in the component</button>

**Comment by pembeci** [comment]
@martypdx, I know and leaving it to the component author makes sense. I was referring to @JonDum's suggestion above.

**Comment by martypdx** [comment]
Hopefully closed by #2095. Not everything in the thread was covered (you can't subscribe DOM events directly, only ractive events from the component), but thinking future issues could be opened to further refine/enhance.
